Переписываю код Playlist Maker с использованием шаблона MVVM.

AlexOtkidach opened the pull request:
2. Перенес логику обработки нажатий кнопок в MainViewModel с использованием LiveData для управления навигацией между экранами.

AlexOtkidach committed:
1. Начал выполнение задания 16 спринта.
2. Перенес логику обработки нажатий кнопок в MainViewModel с
использованием LiveData для управления навигацией между экранами.

AlexOtkidach committed:
Вроде все части задания спринта 16

ArturNurtdinov commented on app/src/main/java/com/example/plm2/presentation/AudioPlayerActivity.kt:
```diff
@@ -72,14 +75,17 @@ class AudioPlayerActivity : AppCompatActivity(), AudioPlayerView {
         if (!isInternetConnected()) {
             // Обработка отсутствия интернет-соединения
-            Toast.makeText(this, "No internet connection", Toast.LENGTH_SHORT).show()
+            Toast.makeText(this, "Нет соединения с Интернетом", Toast.LENGTH_SHORT).show()
```
Все строки необходимо выносить в ресурсы

AlexOtkidach replied:
Поправил

ArturNurtdinov commented on app/src/main/java/com/example/plm2/presentation/AudioPlayerActivity.kt:
```diff
@@ -72,14 +75,17 @@ class AudioPlayerActivity : AppCompatActivity(), AudioPlayerView {
         supportActionBar?.setDisplayShowHomeEnabled(true)
 
         // Проверяем интернет-соединение перед загрузкой треков
         if (!isInternetConnected()) {
```
Обычно UI не проверяет подключение к интернету, а в любом случае делает запрос. Сетевой клиент должен вернуть соответствующую ошибку при отсутствии подключения

ArturNurtdinov commented on app/src/main/java/com/example/plm2/presentation/AudioPlayerActivity.kt:
```diff
@@ -194,4 +211,10 @@ class AudioPlayerActivity : AppCompatActivity(), AudioPlayerView {
-}
+
+    private fun loadTracks() {
+        GlobalScope.launch(Dispatchers.Main) {
```
GlobalScope не рекомендуется использовать в андроид, в активити у тебя есть `lifecycleScope`, во вьюмодели - `viewModelScope`. тебе эту логику необходимо во вьюмодель перенести

AlexOtkidach replied:
Понял

ArturNurtdinov commented on app/src/main/java/com/example/plm2/presentation/AudioPlayerViewModel.kt:
```diff
@@ -0,0 +1,11 @@
+interface AudioPlayerViewModel {
+    val tracksLiveData: LiveData<List<Track>>
+    fun loadTracks()
+}
```
Для вьюмодели можно не делать интерфейс. Интерфейсы нужны для связи слоёв - например интерфейс интерактора для связи presentation и domain. И интерфейс репозитория для связи data - domain

ArturNurtdinov commented on app/src/main/java/com/example/plm2/presentation/MainViewModel.kt:
```diff
@@ -0,0 +1,37 @@
+import androidx.lifecycle.ViewModel
+
+class MainViewModel : ViewModel() {
+    private val _navigateToSearch = MutableLiveData<Boolean>()
```
Экшены (типа навигации и показа тостов/снекбаров) нельзя делать через лайвдату, т.к. это эвенты, которые должны обработаться один раз, лайвдата для этого не подходит. Это называется One Time Event и существует несколько способов реализации такого способа взаимодействия, один из таких примеров - SingleLiveData

ArturNurtdinov commented on app/src/main/java/com/example/plm2/presentation/SearchViewModel.kt:
```diff
@@ -0,0 +1,34 @@
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.viewModelScope
+import com.example.plm2.domain.Track
+import com.example.plm2.data.TracksRepositoryImpl
```
presentation не может зависеть от data слоя, используй подход с объектом `Creator`, о котором рассказывается на курсе, чтобы избежать нарушений архитектуры

ArturNurtdinov reviewed (commented):
UI во многих местах не соответствует макетам - экран аудиоплеера, элемент списка, история поиска, главный экран. Рекомендую поправить эти недочёты сразу

AlexOtkidach committed:
1. Выполнил распределение классов по пакетам

AlexOtkidach committed:
1. Выполнил рефакторинг пакетов,есть ошибки №10 в заметках, в частности с поиском, заняться устранением

AlexOtkidach committed:
1. Устранил проблему с поиском трека

AlexOtkidach committed:
Что нового:
1. В SharedPreferencesManager:
Методы для сохранения и загрузки темы остались без изменений.

2.В aseActivity:
Теперь тема загружается из SharedPreferences и применяется при запуске активности.
3. В SettingsActivity:
При изменении темы, состояние сохраняется и вызывается recreate(), чтобы изменения применились сразу.
4. Вернул правильную логику работы Плейсхолдера
        Проверка соединения с интернетом
        Обработка ошибок и отображение плейсхолдера

AlexOtkidach committed:
Легкие корректировки:

1. Поправил цвет отображения переключателя темной темы в зависимости от выбраной, проверил сохранение темы при сворачивании приложения.

AlexOtkidach committed:
ИСПРАВЛЕНИЕ ОШИБОК 1 из 8
Соблюдается правило зависимостей:
-TrackRepositoryImpl находится в domain слое. Реализация репозитория должна лежать в data слое
-AudioPlayerViewModelImpl зависит от data слоя напрямую (импорт com.example.plm2.data.TracksRepositoryImpl). При этом класс не используется.
-AudioPlayerActivity зависит от data слоя, зависимость - TracksRepositoryImpl
-SearchActivity зависит от data слоя, зависимости - ApiService, TracksRepositoryImpl, SharedPreferencesManager, SearchHistory
-SettingsActivity зависит от data слоя, зависимость - SharedPreferencesManager

AlexOtkidach committed:
ИСПРАВЛЕНИЕ ОШИБОК 2 из 8
❌ Классы Activity отвечают только за отслеживание действий пользователя (нажатия на экран, ввод текста) и за отображение данных на экране:
-SearchActivity напрямую запрашивает данные из SearchHistory для отображения истории поиска, всё взаимодействие должно происходить с помощью вьюмодели

AlexOtkidach committed:
ИСПРАВЛЕНИЕ ОШИБОК 3 из 8
❌ ViewModel зависит только от Interactor или аналогичного интерфейса/класса, объявленного в Domainслое - Не хватает вьюмоделей экрана настроек и плеера

AlexOtkidach committed:
ИСПРАВЛЕНИЕ ОШИБОК 3 из 8
❌ ViewModel зависит только от Interactor или аналогичного интерфейса/класса, объявленного в Domainслое - Не хватает вьюмоделей экрана настроек и плеера

AlexOtkidach committed:
ИСПРАВЛЕНИЕ ОШИБОК 4 из 8
❌ Вся логика по получению или сохранению данных перенесена из Activity в слои Data и Domain - SearchActivity напрямую работает с SearchHistory, а вьюмодель отвеачает только за запрос треков с апи, вьюмодель должна полностью собирать весь стейт активити, а активити только отображать данные по стейту

AlexOtkidach committed:
ИСПРАВЛЕНИЕ ОШИБОК 5 из 8
Теперь:
SharedPreferencesManager - Управление данными SharedPreferences.
PreferencesRepository- Интерфейс для взаимодействия с SharedPreferencesManager.
BaseActivity - Обращается к PreferencesRepository для получения и установки данных SharedPreferences.
SettingsActivity - Работает через ViewModel и Interactor, которые, в свою очередь, обращаются к PreferencesRepository.

AlexOtkidach committed:
ИСПРАВЛЕНИЕ ОШИБОК 7 из 8
❌ Все взаимодействия со слоями Model должны осуществляться через ViewModel:
-SearchViewModel отвечает только за запрос треков с апи, а должна полностью отвечать за всё состояние экрана - историю поиска, прогресс, список треков, обработку ошибок
-AudioPlayerViewModel не используется, т.е. у активити плеера вообще нет вьюмодели
-У SettingsActivity нет вьюмодели

AlexOtkidach committed:
ИСПРАВЛЕНИЕ ОШИБОК:ДИЗАЙН

1. Коррекция дизайна на странице поиска. Стрелка назад. Изменены цвета в темах на верные.
2. Коррекция подсказки в поле Поиск. Изменены цвета в темах на верные.

AlexOtkidach committed:
ИСПРАВЛЕНИЕ ОШИБОК:ДИЗАЙН

1.Востановил цвет курсора снова на синий.

AlexOtkidach committed:
ИСПРАВЛЕНИЕ ОШИБОК:ДИЗАЙН

1.Востановил работу кнопки сброса поля поиск

AlexOtkidach committed:
ИСПРАВЛЕНИЕ ОШИБОК:ДИЗАЙН

1.Исправил цвет преключателя темы

AlexOtkidach committed:
ИСПРАВЛЕНИЕ ОШИБОК 10

AlexOtkidach committed:
Устранил ошибку поиска

AlexOtkidach committed:
Скоректировал работу логики кнопки очистить
